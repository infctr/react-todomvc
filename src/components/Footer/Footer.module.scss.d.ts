export const footer: string;
export const count: string;
export const filters: string;
export const clear: string;
