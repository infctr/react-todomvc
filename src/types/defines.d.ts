declare module 'localStorage' {
  export function setItem(key: string, data: any): void;
  export function getItem(key: string): any;
}
