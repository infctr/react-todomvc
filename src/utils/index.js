export const pluralize = (count, word) => (count === 1 ? word : `${word} s`);

export const storage = (namespace, data) => {
  if (data) {
    try {
      return localStorage.setItem(namespace, JSON.stringify(data));
    } catch (error) {
      return undefined;
    }
  }

  const store = localStorage.getItem(namespace);

  try {
    return (store && JSON.parse(store)) || undefined;
  } catch (error) {
    return undefined;
  }
};

export const switchCase = cases => defaultCase => key =>
  Object.prototype.hasOwnProperty.call(cases, key) ? cases[key] : defaultCase;
