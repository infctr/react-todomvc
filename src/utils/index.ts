export const pluralize = (count: number, word: string) =>
  count === 1 ? word : `${word}s`;

export const storage = (namespace: string, data?: object) => {
  if (data) {
    try {
      return localStorage.setItem(namespace, JSON.stringify(data));
    } catch (error) {
      return undefined;
    }
  }

  const store = localStorage.getItem(namespace);

  try {
    return (store && JSON.parse(store)) || undefined;
  } catch (error) {
    return undefined;
  }
};
