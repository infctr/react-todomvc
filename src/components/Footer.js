import React from 'react';
import { pluralize } from '../utils/index';
import { VisibilityFilters } from '../constants/actionTypes';
import LinkContainer from '../containers/LinkContainer';

const Footer = ({
  count,
  completedCount,
  visibilityFilter,
  onClearCompleted = () => {},
}) => {
  let clearButton = null;
  const activeTodoWord = pluralize(count, 'item');

  if (completedCount > 0) {
    clearButton = (
      <button className="clear-completed" onClick={onClearCompleted}>
        Clear completed
      </button>
    );
  }

  return (
    <footer className="footer">
      <span className="todo-count">
        <strong>{count}</strong> {activeTodoWord} left
      </span>
      <ul className="filters">
        {Object.values(VisibilityFilters).map(filter => (
          <LinkContainer key={filter} filter={filter} />
        ))}
      </ul>
      {clearButton}
    </footer>
  );
};

export default Footer;
