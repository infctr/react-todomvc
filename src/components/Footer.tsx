import React from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/noop';

import { pluralize } from '../utils';
import { VisibilityFilters } from '../types/models';
import FilterLink from './FilterLink';

export interface FooterProps {
  count: number;
  completedCount: number;
  onClearCompleted: () => void;
}

interface DefaultProps {
  readonly onClearCompleted: () => void;
}

const Footer: React.SFC<FooterProps & DefaultProps> = ({
  count,
  completedCount,
  onClearCompleted,
}) => {
  const activeTodoWord = pluralize(count, 'item');

  const clearButton = completedCount > 0 && (
    <button
      type="button"
      className="clear-completed"
      onClick={onClearCompleted}>
      Clear completed
    </button>
  );

  return (
    <footer className="footer">
      <span className="todo-count">
        <strong>{count}</strong> {activeTodoWord} left
      </span>
      <ul className="filters">
        {Object.keys(VisibilityFilters)
          .filter(key => !Number.isNaN(Number(key)))
          .map(key => <FilterLink key={key} filter={Number(key)} />)}
      </ul>
      {clearButton}
    </footer>
  );
};

Footer.propTypes = {
  completedCount: PropTypes.number.isRequired,
  count: PropTypes.number.isRequired,
  onClearCompleted: PropTypes.func,
};

Footer.defaultProps = {
  onClearCompleted: noop,
};

export default Footer;
