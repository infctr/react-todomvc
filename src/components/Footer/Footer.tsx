import React from 'react';
import noop from 'lodash/noop';

import { pluralize } from '../../utils/pluralize';
import { VisibilityFilters } from '../../types/models';

import { FilterLink } from '../FilterLink';

import styles from './Footer.module.scss';

interface IDefaultProps {
  onClearCompleted: () => void;
}

interface IProps extends IDefaultProps {
  count: number;
  completedCount: number;
}

const Footer: React.SFC<IProps> = ({
  count,
  completedCount,
  onClearCompleted,
}) => {
  const activeTodoWord = pluralize(count, 'item');

  const clearButton = completedCount > 0 && (
    <button type="button" className={styles.clear} onClick={onClearCompleted}>
      Clear completed
    </button>
  );

  return (
    <footer className={styles.footer}>
      <span className={styles.count}>
        <strong>{count}</strong> {activeTodoWord} left
      </span>
      <ul className={styles.filters}>
        {Object.keys(VisibilityFilters).map(filter => (
          <FilterLink key={filter} filter={filter as VisibilityFilters} />
        ))}
      </ul>
      {clearButton}
    </footer>
  );
};

// eslint-disable-next-line fp/no-mutation
Footer.defaultProps = {
  onClearCompleted: noop,
};

export default Footer;
