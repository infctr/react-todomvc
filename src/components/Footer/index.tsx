import React, { PureComponent } from 'react';
import noop from 'lodash/noop';

import { pluralize } from 'utils';
import { VisibilityFilters } from 'types/models';

import FilterLink from 'components/FilterLink';

import styles from './index.module.scss';

interface IDefaultProps {
  onClearCompleted: () => void;
}

interface IProps extends Partial<IDefaultProps> {
  count: number;
  completedCount: number;
  onClearCompleted?: () => void;
}

type IPropsWithDefault = IProps & IDefaultProps;

class Footer extends PureComponent<IPropsWithDefault> {
  public static defaultProps: IDefaultProps = {
    onClearCompleted: noop,
  };

  public render() {
    const { count, completedCount, onClearCompleted } = this.props;

    const activeTodoWord = pluralize(count, 'item');

    const clearButton = completedCount > 0 && (
      <button type="button" className={styles.clear} onClick={onClearCompleted}>
        Clear completed
      </button>
    );

    return (
      <footer className={styles.footer}>
        <span className={styles.count}>
          <strong>{count}</strong> {activeTodoWord} left
        </span>
        <ul className={styles.filters}>
          {Object.keys(VisibilityFilters).map((filter: VisibilityFilters) => (
            <FilterLink key={filter} filter={filter} />
          ))}
        </ul>
        {clearButton}
      </footer>
    );
  }
}

export default Footer as React.ComponentClass<IProps>;
