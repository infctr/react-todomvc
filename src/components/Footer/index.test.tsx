import React from 'react';
import { shallow } from 'enzyme';

import { GetComponentProps } from 'utils/types';

import Footer from './index';

describe('Footer component', () => {
  const render = (props: GetComponentProps<typeof Footer>) =>
    shallow(<Footer {...props} />);

  it('should render correctly', () => {
    const tree = render({ completedCount: 5, count: 10 });

    expect(tree).toMatchSnapshot();
  });
});
