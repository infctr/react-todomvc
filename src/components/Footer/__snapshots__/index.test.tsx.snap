// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component should render correctly 1`] = `
<footer
  className="footer"
>
  <span
    className="count"
  >
    <strong>
      10
    </strong>
     
    items
     left
  </span>
  <ul
    className="filters"
  >
    <Connect(FilterLink)
      filter="SHOW_ALL"
      key="SHOW_ALL"
    />
    <Connect(FilterLink)
      filter="SHOW_ACTIVE"
      key="SHOW_ACTIVE"
    />
    <Connect(FilterLink)
      filter="SHOW_COMPLETED"
      key="SHOW_COMPLETED"
    />
  </ul>
  <button
    className="clear"
    onClick={[Function]}
    type="button"
  >
    Clear completed
  </button>
</footer>
`;
